### PR TITLE
lp1769717: Fix decoding of improperly encoded FLAC files

### PR DESCRIPTION
**Pull request by uklotzde**
https://bugs.launchpad.net/mixxx/+bug/1769717

### CODE REVIEW

**Comment by uklotzde** [comment]
Added some commits that hopefully fix the crash during the macOS tests.

**Comment by Be-ing** [comment]
Has anyone reported this issue to the upstream [libFLAC bug tracker](https://sourceforge.net/p/flac/bugs/)?

**Comment by uklotzde** [comment]
No. Maybe we should, but I don't plan to invest even more time for dissecting and explaining this obviously long known issue. This is endless :(

**Comment by uklotzde** [comment]
Already reported?
https://sourceforge.net/p/flac/bugs/446/
https://sourceforge.net/p/flac/bugs/440/

**Comment by Be-ing** [comment]
IIUC, the problem here is in the encoding rather than the decoding, correct? So even if the bug is fixed upstream there will still be people out there with corrupted files for potentially years in the future?

**Comment by uklotzde** [comment]
We have to leave the workaround in our code until the corresponding output sanitizer has been added to libFLAC and the fixed version has been propagated to all of our supported platforms.

**Comment by rryan** [comment]
LGTM. Thanks a ton!

File: src/sources/soundsourceflac.cpp [hunk 1]
```diff
@@ -345,6 +345,25 @@ FLAC__bool SoundSourceFLAC::flacEOF() {
...
+    // Workaround for improperly encoded FLAC files that may contain
+    // garbage in the most significant, unused bits of decoded samples.
+    // Required at least for libFLAC 1.3.2. This workaround might become
+    // obsolete once libFLAC is taking care of these issues internally.
+    // https://bugs.launchpad.net/mixxx/+bug/1769717
+    // https://hydrogenaud.io/index.php/topic,61792.msg559045.html#msg559045
+    FLAC__int32 signBit = static_cast<FLAC__int32>(1) << (bitsPerSample - 1);
+    FLAC__int32 bitMask = (static_cast<FLAC__int32>(1) << bitsPerSample) - 1; // == (signBit << 1) - 1
+    FLAC__int32 maskedSample = decodedSample & bitMask;
+    if (maskedSample & signBit) {
```
Comment (daschuer) [inline src/sources/soundsourceflac.cpp]: This is a relative hot loop.
```
decodedSample <<= 32 - bitsPerSample;
FloatSample = static_cast<CSAMPLE>(decodedSample) / 0xffffffff;
```
Should work fast, without any conditional code.

**Review by daschuer — COMMENTED** [review]
I think we can change the code for no performance lost. Should I issue a PR for that?

File: src/sources/soundsourceflac.cpp (hunk 1 above)
Comment (uklotzde) [reply src/sources/soundsourceflac.cpp]: :+1: I was already thinking if there is some bit magic to improve this workaround. First shifting into the other direction to move out the garbage and then catch the sign bit with a shift into the opposite direction came into my mind. The conditional is really ugly!

Comment (uklotzde) [reply src/sources/soundsourceflac.cpp]: One left shift + float multiplication is sufficient for the entire conversion to CSAMPLE! But I don't get your cast to floating point followed by the masking.

**Comment by uklotzde** [comment]
My proposed solution: #1661

**Pull request merged**